bindings/gumjs/gumv8core.cpp

```diff
@@ -2276,52 +2284,85 @@ gum_v8_native_function_invoke (GumV8NativeFunction * self,
 
-  if (num_args_provided != num_args_required)
+  if ((is_variadic && num_args_provided < num_args_fixed) || (!is_variadic && num_args_provided != num_args_declared))
   {
     _gum_v8_throw_ascii_literal (isolate, "bad argument count");
     return;
   }
 
-  auto rtype = self->cif.rtype;
-
+  auto rtype = cif->rtype;
+  auto atypes = cif->arg_types;
   gsize rsize = MAX (rtype->size, sizeof (gsize));
   gsize ralign = MAX (rtype->alignment, sizeof (gsize));
   auto rvalue = (GumFFIValue *) g_alloca (rsize + ralign - 1);
   rvalue = GUM_ALIGN_POINTER (GumFFIValue *, rvalue, ralign);
 
   void ** avalue;
   guint8 * avalues;
+  ffi_cif tmp_cif;
+  GumFFIValue tmp_value = { 0, };
 
-  if (num_args_required > 0)
+  if (num_args_provided > 0)
   {
-    avalue = (void **) g_alloca (num_args_required * sizeof (void *));
+    avalue = (void **) g_alloca (MAX (num_args_declared, num_args_provided) * sizeof (void *));
+
+    gsize arglist_size = self->arglist_size;
+    if (is_variadic && num_args_provided > num_args_declared)
+    {
+      atypes = g_newa (ffi_type *, num_args_provided);
+
+      memcpy (atypes, cif->arg_types, num_args_declared * sizeof (ffi_type *));
+      for (gsize i = num_args_declared, type_idx = num_args_fixed; i != num_args_provided; i++)
+      {
+        ffi_type * t = cif->arg_types[type_idx];
 
-    gsize arglist_alignment = self->cif.arg_types[0]->alignment;
-    avalues = (guint8 *) g_alloca (self->arglist_size + arglist_alignment - 1);
+        atypes[i] = t;
+        arglist_size = GUM_ALIGN_SIZE (arglist_size, t->alignment);
+        arglist_size += t->size;
+
+        if (++type_idx >= num_args_declared)
+          type_idx = num_args_fixed;
+      }
+
+      cif = &tmp_cif;
+      if (ffi_prep_cif_var (cif, self->abi, num_args_fixed, num_args_provided,
+          rtype, atypes) != FFI_OK)
+        _gum_v8_throw_ascii_literal (isolate,
+            "failed to compile function call interface");
+    }
+
+    gsize arglist_alignment = atypes[0]->alignment;
+    avalues = (guint8 *) g_alloca (arglist_size + arglist_alignment - 1);
     avalues = GUM_ALIGN_POINTER (guint8 *, avalues, arglist_alignment);
 
     /* Prefill with zero to clear high bits of values smaller than a pointer. */
-    memset (avalues, 0, self->arglist_size);
+    memset (avalues, 0, arglist_size);
 
     gsize offset = 0;
-    for (gsize i = 0; i != num_args_required; i++)
+    for (gsize i = 0; i != num_args_provided; i++)
     {
-      auto t = self->cif.arg_types[i];
+      auto t = atypes[i];
 
       offset = GUM_ALIGN_SIZE (offset, t->alignment);
 
       auto v = (GumFFIValue *) (avalues + offset);
 
       if (!gum_v8_value_to_ffi_type (core,
           (argv != nullptr) ? argv[i] : info[i], v, t))
-        return;
+        _gum_v8_throw_ascii_literal (isolate, "invalid argument value");
       avalue[i] = v;
 
       offset += t->size;
     }
+
+    for (gsize i = num_args_provided; i < num_args_declared; i++)
```

Comment:
Loops should use `!=` for the comparison when possible. (More explicit, but also to turn off-by-one errors into infinite loops, which are way easier to detect than subtle misbehavior.)

Reply:
Can not use != at here, because `num_args_provided` maybe bigger than `num_args_declared`. The code's purpose is filling the `avalue` array with an empty value when `num_args_provided < num_args_declared`, and should not work when `num_args_provided >= num_args_declared`.

Reply:
Perfect!